mitxgraders/helpers/calc.py

```diff
@@ -529,7 +531,7 @@ def eval_number(self, parse_result):
         """
         result = float(parse_result[0])
         if len(parse_result) == 2:
-            result *= self.suffixes[parse_result[1]]
```

Comment:
Was there some weirdness with the magic functions to make this work that precluded its use? It's going to make future debugging really confusing if so!

Reply:
I knew you were going to notice that change :).

On master:
```python
import numpy as np
from mitxgraders.helpers.calc import evaluator
result = evaluator('a/2 + a/2', {'a': np.array(8.0)}, {}, {})[0]
print(result) # prints 4.0
```
In-place operations here are problematic for mutable objects. It hadn't been a problem in earlier releases because parse_results only ever held immutable objects (float, int, complex, etc).

So, in fear of this issue I exorcised all in-place operations from calc.py.

I did add a test in `test_calc.py` to try and ensure that we don't reintroduce mutation later on.

Reply:
Hmm, ok. I think `print(variables['a'])` was what you wanted to point me to here. I think we shouldn't be avoiding in-place operations. Instead, we should avoid exposing references to objects that we don't want mutated. In particular, the `eval_var` method (or whatever we called it) should return a copy of a variable, instead of a reference to the original.

Reply:
An alternative is to use [augmented assignment](https://rszalski.github.io/magicmethods/) magic methods to return new objects instead of mutating current ones.

Reply:
No, I did mean `print(results)`, but i actually didn't fully understand what was happening until you made me re-think it. Some logging explains:
```python
eval_product sees: [array(8.0), '/', 2.0]
variables is: {'a': array(8.0)}
eval_product sees: [array(4.0), '/', 2.0]
variables is: {'a': array(4.0)}
eval_sum sees: [array(2.0), '+', array(2.0)]
variables is: {'a': array(2.0)}
4.0
variables is: {'a': array(4.0)}
```

I like the idea of copying the variable values, that seems transparent. That would still require implementing __iadd__, etc MathArray, which is a bit of a pain, but I think mutative versions would actually be OK if we copied the variables.